tests/test_file_writer.py

```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import unittest
+import sys
+import os
+from model_analyzer.output.file_writer import FileWriter
+
+
+class TestFileWriterMethods(unittest.TestCase):
+    def test_write(self):
+        filename = 'test_file'
+        writer = FileWriter(filename=filename)
+
+        # Write test using create if not exist mode
+        writer.write('test', write_mode='w+')
+
+        # open and read file
+        with open(filename, 'r') as f:
+            content = f.read()
+
+        self.assertEqual(content, 'test')
+
+        # redirect stdout and create writer with no filename
+        old_stdout = sys.stdout
+        sys.stdout = open('test_file_stdout', 'w')
```

Comment:
You do not want to actually write out to STD_OUT in a unit test. The unit test could not fail due to reasons beyond the code (out of memory, no space on disk, etc).

In cases like this, you want to use a string buffer in place of stdout. Then have a file pointer to the buffer instead of stdout and set the writer to use that. You may need to create a `TestFileWriter` that overrides the methods that actually write to a file or have it point internally to a buffer.

Then you also want to test the output. Write now, your test only shows that something was written, not what. 

Finally assertion tests should be in separate test cases.

Reply:
So what I'm doing here is writing to a file on disk called `test_file_stdout`. `sys.stdout` is a file pointer to `STDOUT `maintained by the `sys `module. I don't think I'm actually writing to `STDOUT`. 

Also the test currently writes the string `"test"` to these files, and then the asserts check that the string `"test"` is present

Should it be a single assert per test case? I think that will be too many test cases.

Reply:
You should have not have any file IO in a unit test. If possible, write to a string. This string should be compared to a known string constant. Preferably defined in the same test.